static/src/javascripts/lib/manage-ad-free-cookie.ts

```diff
@@ -0,0 +1,117 @@
+ * 2. The user has opted out of targeted advertising
+ * 3. The url parameter noadsaf is set so ads are temporarily disabled
+ */
+export enum AdFreeCookieReasons {
+	ConsentOptOut = 'consent_opt_out',
+	Subscriber = 'subscriber',
+	ForceAdFree = 'force_ad_free',
+}
+/*
+ * Since the cookie is shared between these different use cases, we need to be careful to
+ * only unset it when it's needed for none of the cases.
+ * We keep track of the reasons for setting the ad cookie in an object in local storage
+ * mapping reasons to expiry times.
+ */
+const AD_FREE_COOKIE_REASON_LS = 'gu.ad_free_cookie_reason';
+type AdFreeCookieReasonExpiries = Partial<Record<AdFreeCookieReasons, string>>;
+
+const getAdFreeCookie = (): string | null =>
+	getCookie({ name: AD_FREE_USER_COOKIE });
+
+const adFreeDataIsOld = (): boolean => {
+	const { switches } = window.guardian.config;
+	return (
+		Boolean(switches.adFreeStrictExpiryEnforcement) &&
+		cookieIsExpiredOrMissing(AD_FREE_USER_COOKIE)
+	);
+};
+
+const adFreeDataIsPresent = (): boolean => {
+	const cookieVal = getAdFreeCookie();
+	if (!cookieVal) return false;
+	return !Number.isNaN(parseInt(cookieVal, 10));
+};
+
+const getAdFreeCookieReason = () => {
+	const adFreeReasonString = localStorage.getItem(AD_FREE_COOKIE_REASON_LS);
+	return JSON.parse(adFreeReasonString ?? '{}') as AdFreeCookieReasonExpiries;
+};
+
+/*
+ * Set the ad free cookie and update ad free cookie reason.
+ * Don't set the reason for ForceAdFree as this is supposed to be overridden.
+ *
+ * @param reason
+ * @param daysToLive - number of days the cookie should be valid
+ */
+const setAdFreeCookie = (reason: AdFreeCookieReasons, daysToLive = 1): void => {
+	if (reason !== AdFreeCookieReasons.ForceAdFree) {
+		const adFreeReason = getAdFreeCookieReason();
+		adFreeReason[reason] = timeInDaysFromNow(daysToLive);
+
+		localStorage.setItem(
+			AD_FREE_COOKIE_REASON_LS,
+			JSON.stringify(adFreeReason),
+		);
+	}
+
+	const expires = new Date();
+	expires.setMonth(expires.getMonth() + 6);
+	setCookie({
+		name: AD_FREE_USER_COOKIE,
+		value: expires.getTime().toString(),
+		daysToLive,
+	});
+};
+
+/*
+ * Remove the given ad free reason.
+ * If as a result, all ad free reasons are expired or null, unset the ad free cookie.
+ * @param reason - the reason for which we may want to unset the ad free cookie
+ */
+const maybeUnsetAdFreeCookie = (reason: AdFreeCookieReasons): void => {
+	const adFreeReason = getAdFreeCookieReason();
+
+	delete adFreeReason[reason];
+
+	localStorage.setItem(
+		AD_FREE_COOKIE_REASON_LS,
+		JSON.stringify(adFreeReason),
+	);
+
+	const allExpired = Object.entries(AdFreeCookieReasons).every(
+		([, reason]) => {
+			const expiry = adFreeReason[reason];
+			if (expiry) {
+				const expiryTime = parseInt(expiry, 10);
+				const timeNow = new Date().getTime();
+				return timeNow >= expiryTime;
+			}
+			return true;
+		},
+	);
+
+	if (allExpired) {
+		removeCookie({ name: AD_FREE_USER_COOKIE });
+	}
+};
+
+export {
```

Comment:
Is there any logic here that could benefit from a unit test? Only if you think its worth it!